### PR TITLE
new shape property for list arrays, viewed as sparse matrix

### PR DESCRIPTION
**Pull request by marc9595**
New property for list arrays viewed as sparse matrix. This is to done to still get the shape of the sparse matrix while avoiding casting to `csr_matrix` in `predictive_services_model`.

### CODE REVIEW

File: fletcher/base.py [hunk 1]
```diff
@@ -222,6 +222,27 @@ def __init__(self, array, dtype=None, copy=None):
 
         self._dtype = FletcherDtype(self.data.type)
 
+    @property
+    def shape_as_sparse_matrix(self):
```
Comment (artemru) [inline fletcher/base.py]: why not `Fletcher(self.data.flatten()).max()` ?

Comment (artemru) [reply fletcher/base.py]: flatten does not create a copy !

Comment (artemru) [reply fletcher/base.py]: btw, you can define a method `Fletcher.flatten -> Fletcher`

Comment (artemru) [reply fletcher/base.py]: image

Comment (marc9595) [reply fletcher/base.py]: Done.

Comment (artemru) [reply fletcher/base.py]: instead of defining a new method,`shape_as_sparse_matrix`, I would support `list[primitive_type]` case in `self._reduce`.

Comment (marc9595) [reply fletcher/base.py]: The problem is that it is not clear what the "max" of a list array should be, for example: 
```
In: df = pd.DataFrame({'a': [[1,2], [3,4]]})
In: df['a'].max()
Out: [3,4]
```
So writing it in reduce does not look natural to me.

Comment (artemru) [reply fletcher/base.py]: ok, let's call it `flatten_max`. Or it's even better to create `Fletcher.flatten`  method, and do `fa.flatten().max()`.

Comment (krivonogov) [reply fletcher/base.py]: `fa.flatten().max()` looks nice 👍 @marc9595 wdyt?

File: fletcher/base.py [hunk 2]
```diff
@@ -222,6 +222,14 @@ def __init__(self, array, dtype=None, copy=None):
...
```
Comment (artemru) [inline fletcher/base.py]: no need for `shape_as_sparse_matrix` method, try to stay generic !

Comment (artemru) [reply fletcher/base.py]: the caller can do `shape = (len(self), self.flatten()._reduce("max") + 1)` its-self.

File: tests/test_base.py [hunk 3]
```diff
@@ -290,3 +290,9 @@ def test_eq():
         df_test == [1, 2]
 
     assert "Lengths must match to compare" == str(error_length.value)
+
+
+def test_flatten():
+    test = [[1, 2], [3, 4]]
+    fr_test = fr.FletcherArray(test)
+    npt.assert_array_equal(fr_test.flatten(), fr.FletcherArray([1, 2, 3, 4]))
```
Comment (krivonogov) [inline tests/test_base.py]: can you please add a test with some empty chunks and with completely empty array with many empty chunks?
